quiz/quiz.go

```diff
@@ -0,0 +1,182 @@
+	for {
+		record, err := reader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			fmt.Println("Error reading from file:\t", err)
+			return Quiz{}, err
+		}
+
+		quiz.Questions[questionCount] = NewQuestion(record[0], record[1])
+		questionCount++
+	}
+
+	quiz.TotalQuestions = questionCount
+	quiz.CorrectQuestions = 0
+	quiz.Filename = filename
+
+	return quiz, nil
+}
+
+/*
+	Print prints a quiz to stdout with nicely aligned columns thanks to the tabwriter package. Useful for debugging.
+*/
+func (q Quiz) Print() (err error) {
+	fmt.Println("\nQuiz - ", q.Filename)
+	w := new(tabwriter.Writer)
+	w.Init(os.Stdout, 0, 8, 2, '\t', tabwriter.AlignRight)
+	_, err = fmt.Fprintln(w, "Question\tAnswer\tUser Answer\tCorrect")
+	if err != nil {
+		return err
+	}
+
+	for index, value := range q.Questions {
+		question := q.Questions[index]
+		if question.QuestionText != "" {
+			_, err = fmt.Fprintln(w, value)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	_, err = fmt.Fprintln(w)
+	if err != nil {
+		return err
+	}
+
+	err = w.Flush()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+/*
+	Start begins the user-input portion of the quiz. Each question is presented to the user and they enter answers via
+stdin with "\n" being used to specify the end of an answer.
+*/
+func (q Quiz) Start(timer int) (Quiz, error) {
+	var err error
+
+	c := make(chan string)
+	successString, returnedQuiz := takeAndTimeQuiz(&q, timer, c)
+	if successString == "success" {
+		return *returnedQuiz, err
+	} else {
+		return *returnedQuiz, errors.New("quiz timed out before user completed it")
+	}
+
+}
+
+/*
+	takeAndTimeQuiz proctors the quiz, asking the user questions and saving the responses.
+	A Quiz pointer is used as input and output so we can be sure we're looking at the same quiz and not a copy.
+    timer is the time in seconds to allow the quiz to run for. If the user doesn't finish in time, the rest of the answers
+    are considered incorrect.
+*/
+func takeAndTimeQuiz(q *Quiz, timer int, c chan string) (string, *Quiz) {
+	fmt.Printf("\nYou will have %v seconds to answer %v questions.\nPress [Enter] to Start the quiz.\n",
+		timer, q.TotalQuestions)
+	reader := bufio.NewReader(os.Stdin)
+	_, _ = reader.ReadString('\n')
+
+	// We'll return the first thing down this channel, whether it be the timer result or the quiz result. Either way, a
+	// pointer to the quiz will be returned so that partial credit can be given.
+	ch := make(chan string)
+	ch = c
+
+	// start the timer
+	fmt.Printf("Starting timer for %v seconds.\n", timer)
+	quizTimer := time.NewTimer(time.Duration(timer) * time.Second)
+	defer quizTimer.Stop()
+
+	// this goroutine waits on the timer and sends a "timeout" string down the channel when it is expired.
+	// takeAndTimeQuiz returns the first thing down the channel, so the timer will 'interrupt' the quiz if it's still going
+	go func() {
+		<-quizTimer.C
+		fmt.Println("\nTimer has run out!")
+		ch <- "timeout"
+	}()
```

Comment:
You might be able to get away with using this
https://golang.org/pkg/time/#AfterFunc